### PR TITLE
Improve NativeHandle debuggability [ECR-2528]

### PR DESCRIPTION
**Pull request by DENMROOT**
## Overview
NativeHandle preserve the original value of the handle after invalidation.

---
See: https://jira.bf.local/browse/ECR-2528

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has Javadoc
- [x] Method preconditions are checked and documented in the Javadoc of the method
- [x] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

### CODE REVIEW

**Comment by MakarovS** [comment]
@DENMROOT You probably wanted to appoint reviewers, not assignees?

**Comment by vitvakatu** [comment]
Funny, I'm 100% sure Github did not support multiple assignees before.

File: exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java [hunk 1]
```diff
@@ -66,11 +69,11 @@ private void checkValid() {
    * Returns true if this native handle is valid.
    */
   final boolean isValid() {
-    return nativeHandle != INVALID_NATIVE_HANDLE;
+    return nativeHandle != INVALID_NATIVE_HANDLE && isValid;
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: I wonder if we shall _allow_ accessing `nullptr` == 0, just to simplify the condition here? @vitvakatu , @DarkEld3r , we can't have `nullptr` as a handle anywhere, can we?

File: exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java [hunk 2]
```diff
@@ -53,6 +51,7 @@ void close() {
 
     nativeHandle.close();
     assertFalse(nativeHandle.isValid());
+    assertThat(nativeHandle.toString()).contains(convertHandle(handle));
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java]: Nit: I'd suggest using a string so that the expected format is apparent and this code is more readable: "11" or "0x11" — whatever it is.

**Comment by dmitry-timofeev** [comment]
> Funny, I'm 100% sure Github did not support multiple assignees before.

I saw it work a couple of months ago, I believe.

File: exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java (hunk 1 above)
Comment (vitvakatu) [reply exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: I believe we can safely use `return isValid` here, yep. We will be able to test debugging capabilities in case of any... hmmm.... troubles :D

Comment (DENMROOT) [reply exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: And what about creating incorrect handles from constructor ? 
new NativeHandel(INVALID_NATIVE_HANDLE) - do we allow this ?

File: exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java (hunk 2 above)
Comment (DENMROOT) [reply exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java]: I'll create String constant for this.

File: exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java [hunk 3]
```diff
@@ -23,53 +23,53 @@
...
 import org.junit.jupiter.api.Test;
 
-
 class NativeHandleTest {
 
   private NativeHandle nativeHandle;
 
+  private static long HANDLE = 0x11L;
+
+  private static final String NATIVE_HANDLE = "11";
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java]: I can't say it is better, it is not the same constant that you reuse in code because it *must* have the same value. It just _happens_ that tests use the same value, therefore, extracting these constants does not make them clearer, on contrary — it makes them harder to understand, because you can't immediately see that the second is another representation of the first.

On top of that, the naming does not help either.

File: exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java (hunk 1 above)
Comment (vitvakatu) [reply exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: I'm not completely sure, but we can allow it

Comment (DENMROOT) [reply exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: @dmitry-timofeev Allow creation of a handle from INVALID_NATIVE_HANDLE ? Should it be VALID until we invalidate it manually ?

Comment (dmitry-timofeev) [reply exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: I'd go with a stricter rule first and relax later if needed: forbid instantiation, if it *ever* (which we currently doubt, if I understand correctly) need that, allow.

File: exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java [hunk 4]
```diff
@@ -23,53 +23,53 @@
...
 
-
 class NativeHandleTest {
 
   private NativeHandle nativeHandle;
 
+  private static long HANDLE = 0x11L;
+
+  private static final String HANDLE_STRING_REPRESENTATION =
+      Long.toHexString(HANDLE).toUpperCase();
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java]: "11";

File: exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java (hunk 1 above)
Comment (dmitry-timofeev) [reply exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: The idea was to simplify this method — i.e., remove `nativeHandle != INVALID_NATIVE_HANDLE` since this condition is not possible because it is enforced on the instantiation.

File: exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java [hunk 5]
```diff
@@ -21,7 +21,7 @@
 import com.google.common.base.MoreObjects;
 
 /**
- * An implementation-specific handle to the native object. Once closed, can no longer be accessed.
+ * An implementation-specific handle to the native object.
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: Why this change if this is still true?

File: exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/AbstractCloseableNativeProxyTest.java [hunk 6]
```diff
@@ -126,9 +113,8 @@ void getNativeHandle_ShallFailIfProxyIsClosed() {
   @Test
   void getNativeHandle_ShallFailIfInvalid() {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/AbstractCloseableNativeProxyTest.java]: The test name is no longer correct.

File: exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java (hunk 4 above)
Comment (DENMROOT) [reply exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java]: > I can't say it is better, it is not the same constant that you reuse in code because it must have the same value. It just happens that tests use the same value, therefore, extracting these constants does not make them clearer, on contrary — it makes them harder to understand, because you can't immediately see that the second is another representation of the first.
- No ?

File: exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java [hunk 7]
```diff
@@ -40,10 +40,10 @@
    * @throws IllegalStateException if this native handle is invalid (nullptr)
    */
   public NativeHandle(long nativeHandle) {
+    checkState(nativeHandle != INVALID_NATIVE_HANDLE, "This handle is not valid: %s", this);
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: Do not invoke toString on a partially constructed object

File: exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/AbstractCloseableNativeProxyTest.java [hunk 8]
```diff
@@ -111,7 +110,7 @@ void getNativeHandle_ShallFailIfProxyIsClosed() {
   }
 
   @Test
-  void getNativeHandle_ShallFailIfInvalid() {
+  void createNativeHandle_ShallFailIfInvalid() {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/AbstractCloseableNativeProxyTest.java]: NativeProxy, it is not a test of a handle

File: exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java (hunk 4 above)
Comment (dmitry-timofeev) [reply exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java]: I am not sure I understand what you mean. No to what and why?

File: exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java [hunk 9]
```diff
@@ -53,24 +63,21 @@ public long get() {
 
   @Override
   public void close() {
-    if (isValid()) {
+    if (isValid) {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: I'd consider keeping the method calls as more refactoring-friendly (if we change what isValid means)

File: exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java (hunk 7 above)
Comment (DENMROOT) [reply exonum-java-binding-core/src/main/java/com/exonum/binding/proxy/NativeHandle.java]: Good catch , thanks

File: exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java (hunk 4 above)
Comment (DENMROOT) [reply exonum-java-binding-core/src/test/java/com/exonum/binding/proxy/NativeHandleTest.java]: Trying to make HANDLE_STRING_REPRESENTATION more clear.
